proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+  we also include any bundle that also controls the warning.
+
+Detailed specification
+~~~~~~~~~~~~~~~~~~~~~~
+
+The text above defines language editions generally. This section
+instantiates the general design with some specifics. These language
+editions affect language features (as described by extension name or
+warning flag in the chart below), as well as error messages (not described
+in any detail). They do not affect other aspects of GHC.
+
+The chart below classifies all current extensions and warnings.
+
+* The GHC2021 column has an X for extensions enabled as part of GHC2021.
+  For warnings, it lists if the warning is part of a current warning bucket.
+
+* The next four columns describe ``Student2024``, ``Stable2024``, ``Experimental2024``, and ``Latest``,
+  respectively. Here is the key for these columns:
+
+  * Y: an extension is turned on
+
+  * blank: the feature is available, but not on
+
+  * N: trying to enable the feature is an error
+
+  * W: the warning warns by default
+
+  * E: the warning is an error by default
+
+  * W!: the warning warns by default and cannot be turned off
+
+  * E!: the warning is an error by default and cannot be turned off or made into a warning
+
+* The next several columns describe semantic bundles of options.
+
+  * ``FancyTypes``: The user should enable this if they want fancy types. By enabling
+    this semantic bundle, the user takes responsibility for understanding type inference
+    and interactions at a deeper level. In an enterprise setting, the user takes the
+    responsibility for training future collaborators in these techniques.
+
+    Includes the following: ``DataKinds``, ``ExistentialQuantification``, ``GADTs``,
+    ``GADTSyntax``, ``ImpredicativeTypes`` (allowing polytypes in visible type applications
+    and in type parameters, no Quick Look), ``QuantifiedConstraints``, ``RankNTypes``,
+    ``RequiredTypeArguments``, ``RoleAnnotations``, ``TypeAbstractions``, ``TypeData``,
+    ``TypeFamilies``, ``-Wterm-variable-capture``.
+
+  * ``DoSyntax``: This enables extra syntactic support around ``do``-notation. Someone
+    who enables this option takes responsibility for understanding the extra syntax and
+    for training others to understand that syntax.
+
+    Includes the following: ``Arrows``, ``QualifiedDo``, ``RecursiveDo``
+
+  * ``Classic``: In some places, Haskell has recently evolved, and the recommendation of
+    the new language editions are to use the new syntax or features. In a few places, users
+    can enable ``Classic`` mode to use older-style Haskell. A user enabling this feature
+    takes responsibility for keeping abreast with today's Haskell best practices and determining
+    when it is appropriate to migrate to the new style.
+
+    Includes the following: ``FieldSelectors``, ``NoPolyKinds``, ``StarIsType``, ``-Wno-deriving-typeable``,
+    ``-Wno-prepositive-qualified-module``, ``-Wno-type-equality-out-of-scope``
```

Comment:
I think it's interesting that this is the only bundle that includes both extensions and warning settings. If it were not for this I would be inclined to say that maybe what we want is _warning groups_, e.g. so you could write `-Werror=incomplete` and toggle all the errors in the "incomplete warning group" into errors. I still think that's a plausible alternative, though - if `Classic` is the _only_ example of such a bundle then maybe it would be cleaner to have extension groups and warning groups separately?

Reply:
Maybe. But I continue to think that the division between language extension and warning is very arbitrary, and so separating out these two from each other seems unhelpful, to me.